File: nvpl_blas/f77/saxpby.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     SAXPBY 
!
!******************************************************************************
      program   SAXPBY_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      integer         n, incx, incy, i
      parameter       (n=5, incx=1, incy=1)
      real            alpha, beta
      parameter       (alpha=1.0, beta=2.0)
      integer         xsize, ysize
      parameter       (xsize=1+(n-1)*abs(incx),
     &                 ysize=1+(n-1)*abs(incy))
      real            x(xsize), y(ysize)
      data            (x(i), i=1,xsize)/1.0,-3.7,4.3,0.0,5.0/,
     &                (y(i), i=1,ysize)/ysize*2.0/

*       External Subroutines
      external        print_svector 
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx, incy, alpha, beta
      call print_svector(x,n,incx,'X')
      call print_svector(y,n,incy,'Y')
*
*      Call SAXPBY subroutine
      call SAXPBY(n,alpha,x,incx,beta,y,incy)
*
      print*
      print 101
      call print_svector(y,n,incy,'Y')

      stop

 99   format('Example: SAXPBY scales vector X and vector Y by two'
     &       ' scalars, adds them to one another, and stores in the'
     &       ' vector Y')
 100  format('#### args: n=',i1,', incx=',i1,', incy=',i1,', '
     &       'alpha=',f3.1,', beta=',f3.1)
 101  format('The result of SAXPBY: ')
      end

File: nvpl_blas/c/saxpby.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_saxpby
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 5;
    float * X;
    float * Y;
    float alpha = 1.0f;
    float beta = 0.5f;
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 1;

    printf("\nExample: cblas_saxpby for computing y := alpha * x + beta * y\n");
    printf("#### args: n=%" PRId64 ", incx=%" PRId64 ", incy=%" PRId64 ", alpha=%f, beta=%f\n", 
            (int64_t)N, (int64_t)incX, (int64_t)incY, alpha, beta);

    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (N - 1) * labs(incY);
    // allocate memory
    X = (float *)malloc(len_x * sizeof(float));
    Y = (float *)malloc(len_y * sizeof(float));

    // fill data
    fill_svector(X, N, incX);
    fill_svector(Y, N, incY);

    // print input data
    print_svector(X, N, incX, "X");
    print_svector(Y, N, incY, "Y");

    // call cblas_saxpby
    cblas_saxpby(N, alpha, X, incX, beta, Y, incY);

    // print result
    printf("\nThe result of cblas_saxpby:\n");
    print_svector(Y, N, incY, "Y");

    // release memory
    free(X);
    free(Y);
    return 0;
}
